module/x/gravity/migrations/v2/store.go

```diff
@@ -0,0 +1,20 @@
+package v2
+
+import (
+	sdktypes "github.com/cosmos/cosmos-sdk/types"
+	paramstypes "github.com/cosmos/cosmos-sdk/x/params/types"
+	"github.com/peggyjv/gravity-bridge/module/v3/x/gravity/types"
+)
+
+func MigrateParamStore(ctx sdktypes.Context, subspace paramstypes.Subspace) error {
+	if subspace.HasKeyTable() {
```

Comment:
Why are we manually manipulating the params subspace? You can pass the keeper to the migration function and set the parameters.

Reply:
The example in the sdk staking module did it this way; the `paramSpace` field is private so I pass it in directly from the keeper:

```go
// Migrate2to3 migrates from consensus version 2 to 3.
func (m Migrator) Migrate2to3(ctx sdk.Context) error {
	return v2.MigrateParamStore(ctx, m.keeper.paramSpace)
}
```

Is the concern it's acting on a copy of the paramSpace and not a reference of the one in the keeper?

Reply:
I'm not totally sure how acting directly on the subspace would behave. To the point above though, what's going to happen is the upgrade handler function will run (which is where anyone moving to the new version of the module will be putting their custom code) and then `RunMigrations` will be called which will call `Migrate2to3`. If someone wants specific values for new params, they will have to specify them in the upgrade handler, but they will be overwritten by the defaults here. What I was getting at with the keeper is that there should already be functions for setting params in there -- we can use the keeper function to read the params, inspect/modify the values, and then write the params, all using the types and methods we've already defined, rather than messing around with subspace internals.

Reply:
We're adding keys to the param set table, I didn't think you could do that with the keeper? Will look into it.

Reply:
The gravity keeper implements a method for using its `Params` type to read and set the values in the store. It's how e.g. the genesis value read/writes work.

https://github.com/PeggyJV/gravity-bridge/blob/76006da2934242631ef1fefe7e2112f26ef2b7da/module/x/gravity/keeper/keeper.go#L390-L403

Reply:
Whether or not it will marshal missing values into zero values I'm not sure. If so, you could probably just do it using the v3 code. If not, you'd have to use a definition from v2 in a v2 migrations package (like the deleted types file) to read it, and then translate it into the new params type from v3 to write it. That might be kinda weird and complicated in that case, and based on the keeper's implementations of `paramSpace` manipulation you could in theory also set it directly, but we still have the "overwriting with defaults" issue mentioned earlier.

Reply:
`GetParams` uses `Subspace.GetParamSet()`, which calls `Subspace.Get` on each item returned by `types.ParamSetPair()`. The new `ParamSetPair` will return two keys that don't exist in the store, so when `Get` is called on either of them it will panic.

The other issue I discovered while trying to confirm this behavior is that `migrations/v2` can't import the keeper package because keeper already imports v2 in `modules.go`

Reply:
Ah, I forgot the migrations package relies on doing a bunch of direct store manipulation. Maybe you will have to edit it directly after all.

Reply:
Further I don't understand what you mean about "someone want[ing] specific values for params"? They are global values set by governance, so it's safe to set them as a default value. I'm missing your point I think.

Reply:
Oh you're saying because this is a module that can be consumed elsewhere, nvm